plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/Fabric8DelegatedOAuthAPI.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2016-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package com.redhat.che.multitenant;
+
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+import com.redhat.che.multitenant.Fabric8AuthServiceClient.GithubToken;
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+import org.eclipse.che.api.auth.shared.dto.OAuthToken;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.dto.server.DtoFactory;
+import org.eclipse.che.multiuser.keycloak.server.oauth2.DelegatedOAuthAPI;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author Angel Misevski (amisevsk@redhat.com) */
+public class Fabric8DelegatedOAuthAPI extends DelegatedOAuthAPI {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Fabric8DelegatedOAuthAPI.class);
+  private final Fabric8AuthServiceClient authServiceClient;
+
+  @Inject
+  public Fabric8DelegatedOAuthAPI(Fabric8AuthServiceClient authServiceClient) {
+    super(authServiceClient);
+    this.authServiceClient = authServiceClient;
+  }
+
+  @Override
+  public Response authenticate(
+      UriInfo uriInfo,
+      String oauthProvider,
+      List<String> scopes,
+      String redirectAfterLogin,
+      HttpServletRequest request)
+      throws BadRequestException {
+    String accountLinkingUrl =
+        authServiceClient.getAccountLinkingURL(null, oauthProvider, redirectAfterLogin);
+    if (Strings.isNullOrEmpty(accountLinkingUrl)) {
+      LOG.error("Failed to get GitHub account linking URL");
+      return Response.status(500).build();
+    }
+    return Response.temporaryRedirect(URI.create(accountLinkingUrl)).build();
+  }
+
+  @Override
+  public OAuthToken getToken(String oauthProvider)
+      throws ForbiddenException, BadRequestException, NotFoundException, ServerException,
+          UnauthorizedException {
+    try {
+      GithubToken token = authServiceClient.getGithubToken();
```

Comment:
Don't you think it would be better to get the `tokenEndpoint` and `linkEndpoint` from `fabric8.*` properties whose name would derive from the `oauthProvider` name ? That would make it much more generic instead of having hard-wired for the GitHub provider ?

Reply:
I agree that we should have properties for the endpoints, but as we discussed above, I think cleaning up the properties used in rh-che should happen first. There are already 3-4 properties that do similar things but don't seem to be used. Before this PR, the `che.fabric8.auth.endpoint` property wasn't used anywhere at all, and setting it did nothing. 

E.g. the oidc provider property is used direclty in `OpenShiftUserTokenProvider`, but is ignored in [`KeycloakServiceClient`](https://github.com/eclipse/che/blob/a59db0a176a82387c78b76201d42050960acced6/multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakServiceClient.java#L103). The documentation around the different properties and how they are used is pretty unclear, so we should take a separate issue to clean up properties and make everything consistent.

Reply:
It is used / checked in the [KeycloakSettings](https://github.com/eclipse/che/blob/a59db0a176a82387c78b76201d42050960acced6/multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakSettings.java) class though. But it seems that `KeycloakServiceClient` was implemented after the OIDC provider support change, but without taking it into account, which explains why the `OIDC_PROVIDER_SETTING` property wasn't used.